src/kokkos_abstraction.hpp

```diff
@@ -591,6 +597,65 @@ inline void par_for_outer(OuterLoopPatternTeams, const std::string &name,
       });
 }
 
+// Inner parallel loop using TeamThreadRange
+template <typename Function>
+KOKKOS_INLINE_FUNCTION void
+par_for_inner(InnerLoopPatternTTR, team_mbr_t team_member, const int nl, const int nu,
+              const int kl, const int ku, const int jl, const int ju, const int il,
+              const int iu, const Function &function) {
+  const int Nn = nu - nl + 1;
+  const int Nk = ku - kl + 1;
+  const int Nj = ju - jl + 1;
+  const int Ni = iu - il + 1;
+  const int NjNi = Nj * Ni;
+  const int NkNjNi = Nk * NjNi;
+  const int NnNkNjNi = Nn * NkNjNi;
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team_member, NnNkNjNi),
+                       [&](const int &idx) {
+                         int n = idx / NkNjNi;
+                         int k = (idx - n * NkNjNi) / NjNi;
+                         int j = (idx - n * NkNjNi - k * NjNi) / Ni;
+                         int i = idx - n * NkNjNi - k * NjNi - j * Ni;
+                         n += nl;
+                         k += kl;
+                         j += jl;
+                         i += il;
+                         function(n, k, j, i);
+                       });
+}
+template <typename Function>
+KOKKOS_INLINE_FUNCTION void par_for_inner(InnerLoopPatternTTR, team_mbr_t team_member,
+                                          const int kl, const int ku, const int jl,
+                                          const int ju, const int il, const int iu,
+                                          const Function &function) {
+  const int Nk = ku - kl + 1;
+  const int Nj = ju - jl + 1;
+  const int Ni = iu - il + 1;
+  const int NkNjNi = Nk * Nj * Ni;
+  const int NjNi = Nj * Ni;
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team_member, NkNjNi), [&](const int &idx) {
+    int k = idx / NjNi;
+    int j = (idx - k * NjNi) / Ni;
+    int i = idx - k * NjNi - j * Ni;
+    k += kl;
+    j += jl;
+    i += il;
+    function(k, j, i);
+  });
+}
+template <typename Function>
+KOKKOS_INLINE_FUNCTION void par_for_inner(InnerLoopPatternTTR, team_mbr_t team_member,
+                                          const int jl, const int ju, const int il,
+                                          const int iu, const Function &function) {
+  const int Nj = ju - jl + 1;
+  const int Ni = iu - il + 1;
+  const int NjNi = Nj * Ni;
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team_member, NjNi), [&](const int &idx) {
+    int j = idx / Ni + jl;
+    int i = idx % Ni + il;
```

Comment:
Not sure if it will make a difference, but modulo could be slow on GPU. But integer multiplication is also not the fastest...

Reply:
Yeah that's worth looking at. I didn't think too hard about this.